### PR TITLE
Owner and Property APIs

### PR DESCRIPTION
**Pull request by Samrudhsaish**
Review my code and let me know the modifications.

### CODE REVIEW

File: src/main/java/com/acro/dev/propmgnt/entity/Property.java
Comment (Ranianirudh) [inline src/main/java/com/acro/dev/propmgnt/entity/Property.java:1]: If you add the relationship properties like CascadeType.All, then we can solve your error in Delete endpoint.

File: src/main/java/com/acro/dev/propmgnt/entity/Owner.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        @Column(name="owner_phonenumber")
+        private Long ownerPhoneNumber;
+
+        @Column(name="ein_number")
+        private Long einNumber;
+
+         @OneToMany(mappedBy = "owner")//1o----many properties
+         //@JoinColumn(name="property_id")
+         List<Property> PropertyList=new ArrayList<>();
```
Comment (Ranianirudh) [inline src/main/java/com/acro/dev/propmgnt/entity/Owner.java]: Initialization is not needed here(ArrayList).

File: pom.xml [hunk 2]
```diff
@@ -17,6 +17,12 @@
 		<java.version>17</java.version>
 	</properties>
 	<dependencies>
+		<dependency>
+			<groupId>com.mysql</groupId>
+			<artifactId>mysql-connector-j</artifactId>
```
Comment (rambabu) [inline pom.xml]: dependency are not in Order

File: src/main/java/com/acro/dev/propmgnt/advice/HandlingErrors.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+@RestControllerAdvice
+public class HandlingErrors {
+    private static final Logger LOGGER= LoggerFactory.getLogger(OwnerController.class);
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/advice/HandlingErrors.java]: why logger has OwnerController.classs it should be HandlingErrors.class

File: src/main/java/com/acro/dev/propmgnt/advice/HandlingErrors.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            errorMap.put(error.getField(), error.getDefaultMessage());
+        });
+        LOGGER.warn("Received BadRequest");
+        return errorMap;
+    }
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(Exception.class)
+    public Map<String, String> handleError(Exception e) {
+        LOGGER.error("Resource not found", e);  // logger
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/advice/HandlingErrors.java]: this method not fully implemented ?? But you can work on it later

File: src/main/java/com/acro/dev/propmgnt/config/SchedulingConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+@Configuration
+@EnableScheduling
+public class SchedulingConfiguration {
+    @Scheduled(cron="0 05 10 15 * ?")
+    public void ScheduleTaskUsingCronExpression(){
+        long now=System.currentTimeMillis()/1000;
+        System.out.println("Schedule tasks using cron jobs " + now);
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/config/SchedulingConfiguration.java]: Remove SOP and use logger

File: src/main/java/com/acro/dev/propmgnt/config/SchedulingConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.acro.dev.propmgnt.config;
+
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+@Configuration
+@EnableScheduling
+public class SchedulingConfiguration {
+    @Scheduled(cron="0 05 10 15 * ?")
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/config/SchedulingConfiguration.java:13]: read cron expression from Application.properties
@Scheduled(cron = "${cron.expression}")
 @Samrudhsaish @Ranianirudh 
 https://www.petrikainulainen.net/programming/spring-framework/spring-from-the-trenches-using-environment-specific-cron-expressions-with-the-scheduled-annotation/

File: src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import jakarta.validation.Valid;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("api/owner")
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java:17]: Read path from Const file.
ex create a class RestConstants,java and define a string , use it here

Comment (rambabu) [reply src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java:17]: Restcosnbtants {
	
	String static final owner_api_base_path = "api/owner";

}

File: src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    public OwnerController(@Autowired OwnerService ownerService) {
+
+        this.ownerService = ownerService;
+    }
+
+
+    @PostMapping("/")
+    public ResponseEntity<OwnerResponse> createOwner(@RequestBody @Valid OwnerRequest ownerRequest) {
+        if (ownerRequest != null) {
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java]: okay with not null but you expore more how to handle @notnull validation in Method level

File: src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<OwnerResponse> updateOwner(@PathVariable Long id,
+                                                     @RequestBody @Valid OwnerRequest ownerRequest) {
+        if (ownerRequest != null) {
+            OwnerResponse ownerResponse = ownerService.updateOwner(id, ownerRequest);
+            if (ownerResponse != null)
+                return ResponseEntity.ok(ownerResponse);
+        }
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java:55]: need to have clean Code

File: src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<OwnerResponse> findOwnerById(@PathVariable Long id) {
+        OwnerResponse ownerResponse = ownerService.findOwnerById(id);
+        if (ownerResponse != null)
+            return ResponseEntity.ok(ownerResponse);
+
+        LOGGER.error("Invalid ownerId");
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java]: return 404

File: src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        if (owners != null)
+            return ResponseEntity.ok(owners);
+        LOGGER.info("Failed to receive list of owners");
+        return ResponseEntity.ok(Collections.emptyList());
+    }*/
+
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Boolean> deleteOwnerById(@PathVariable Long id) {
+        boolean res = this.ownerService.deleteOwnerById(id);
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java]: before deleting check he is present or not if not present return 404

File: src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        LOGGER.error("Deleted Successfully");
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/OwnerController.java]: it should be Logger.info

File: src/main/java/com/acro/dev/propmgnt/controller/PropertyController.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import jakarta.validation.Valid;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("api/property")
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/controller/PropertyController.java]: follow Above Code review Comments

File: src/main/java/com/acro/dev/propmgnt/entity/Address.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package com.acro.dev.propmgnt.entity;
+
+import jakarta.persistence.*;
+import lombok.Data;
+
+//@EqualsAndHashCode(callSuper = true)
+@Entity
+@Data
+@Table(name="address")
+public class Address extends BaseEntity {
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/entity/Address.java:11]: override equals and Hashcode

File: src/main/java/com/acro/dev/propmgnt/entity/Owner.java (hunk 1 above)
Comment (rambabu) [reply src/main/java/com/acro/dev/propmgnt/entity/Owner.java]: i see missing 1-M address relation

File: src/main/java/com/acro/dev/propmgnt/entity/Owner.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+        //owner ----many properties(in propert table have ownerid)
+
+        //@OneToMany(mappedBy = "owner")//1o----many repairs
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/entity/Owner.java]: clean the unnessary code

File: src/main/java/com/acro/dev/propmgnt/request/AddressRequest.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+public class AddressRequest {
+   // @NotNull
+    private Long addressId;
+
+    @NotEmpty
+    private String lineOne;
+    @NotEmpty
+    private String lineTwo;
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/request/AddressRequest.java:18]: linetwo address can be empty incase of SINGLE FAMILY HOME

File: src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+public class OwnerServiceImpl implements  OwnerService {
+
+    private final OwnerRepository ownerRepository;
+    private static final Logger LOGGER= LoggerFactory.getLogger(OwnerController.class);
+    public OwnerServiceImpl(@Autowired OwnerRepository ownerRepository) {
+        this.ownerRepository = ownerRepository;
+    }
+
+    //Owner owner = new Owner();
+    // EmployeeIdentificationNumber ein=new EmployeeIdentificationNumber();
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: REMOVE DEAD CODE

File: src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+        owner.setEinNumber(ownerRequest.getEinNumber());
+        owner = ownerRepository.save(owner);
+        /*OwnerResponse ownerResponse = new OwnerResponse();
+        ownerResponse.setOwnerId(owner.getOwnerId());
+        ownerResponse.setOwnerFirstName(owner.getOwnerFirstName());
+        ownerResponse.setOwnerLastName(owner.getOwnerLastName());
+        ownerResponse.setOwnerEmail(owner.getOwnerEmail());
+        ownerResponse.setOwnerPhoneNumber(ownerRequest.getOwnerPhoneNumber());*/
+        //convert owner to ownerrespon
+        LOGGER.info("Saved Successfully");
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: try to log at least basic information like ID and name of owner in logs

Comment (rambabu) [reply src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: LOGGER.error("Failed to save Owner  name {} ", ownderRequest.getName());

Comment (rambabu) [reply src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: LOGGER.info("Saved Successfully of Owner {} and ID {}",name,id);

File: src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+            existingOwner.setOwnerFirstName(ownerRequest.getOwnerFirstName());
+            existingOwner.setOwnerLastName(ownerRequest.getOwnerLastName());
+            existingOwner.setOwnerEmail(ownerRequest.getOwnerEmail());
+            existingOwner.setOwnerPhoneNumber(ownerRequest.getOwnerPhoneNumber());
+            //existingOwner.setPropertyId(ownerRequest.getPropertyId());
+            Owner updated = ownerRepository.save(existingOwner);
+            LOGGER.info("Updated Successfully");
+            return getOwnerResponse(updated);
+        }
+        return null;
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: why return null ??

File: src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    private OwnerResponse getOwnerResponse(Owner owner) {
+        OwnerResponse ownerResponse = new OwnerResponse();
+        ownerResponse.setOwnerId(owner.getId());
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: if you have more common methods , then move some class like EntityToResponseTransformer.java

File: src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+            Owner ownerOne=owner.get();
+            return getOwnerResponse(ownerOne);
+        }
+          return null;
+    }*/
+
+    @Override
+    public boolean deleteOwnerById(Long id) {
+        try{
+            ownerRepository.findById(id);
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/OwnerServiceImpl.java]: Not a Good of doing this.
if(objedtFound) {
.. do your logic 
} else {
  throw new EmpnOTfoundException (....)
}

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    public PropertyServiceImpl(@Autowired PropertyRepository propertyRepository,
+                               @Autowired OwnerRepository ownerRepository,
+                               @Autowired AddressRepository addressRepository) {
+        this.propertyRepository = propertyRepository;
+        this.ownerRepository = ownerRepository;
+        this.addressRepository = addressRepository;
+
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PropertyController.class);
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: why logger is Controller here , Please use PropertyServiceImpl. i hope it is Copy & Paste Mistake

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+
+    @Override
+    public PropertyResponse createProperty(PropertyRequest propertyRequest) {
+        LOGGER.info("Received request to property");
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java:45]: improve logging statemnets for what property we are logging

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+        Property property = new Property();
+        Address address = new Address();
+        AddressRequest addressRequest = new AddressRequest();
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: why addressRequestNeeded here?
@Samrudhsaish

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+        Optional<Owner> owner = ownerRepository.findById(propertyRequest.getOwnerId());
+        if (owner.isPresent()) {
+            Owner ownerOne = owner.get();
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: why con't you create Address inside if , 
you have initialized Outside

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+
+            Property propertyOne = propertyRepository.save(property);
+            LOGGER.info("Owner address saved");
+            LOGGER.info("Create property");
+            // return getAddressResponse(address);
+            return getPropertyResponse(propertyOne);
+
+
+        } else {
+            return null;
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: throw Exception

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+    }
+
+
+    public PropertyResponse getPropertyResponse(Property property) {
+        PropertyResponse propertyResponse = new PropertyResponse();
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: same like Above Ctrl

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+            existingProperty.setKitchenDimension(propertyRequest.getKitchenDimension());
+            existingProperty.setBedRoomDimension(propertyRequest.getBedRoomDimension());
+            existingProperty.setRent(propertyRequest.getRent());
+
+
+            Property updatedProperty = propertyRepository.save(existingProperty);
+            LOGGER.info("Property Updated Successfully");
+            return getPropertyResponse(updatedProperty);
+        }
+        return null;
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: throw  new CustomException

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+
+
+    @Override
+    public PropertyResponse findPropertyById(Long id) {
+        Optional<Property> property = propertyRepository.findById(id);
+        if (property.isPresent()) {
+            Property propertyOne = property.get();
+            return getPropertyResponse(propertyOne);
+        }
+        return null;
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: same like bbove

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+
+   /* @Override
+    public List<PropertyResponse> getPropertyByOwnerId(Long id) {
+        return null;
+    }*/
+
+    @Override
+    public boolean deletePropertyById(Long id) {
+        try {
+            propertyRepository.findById(id);
```
Comment (rambabu) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: you are just calling find not checking it is  present or not?
pls correct it

File: src/main/java/com/acro/dev/propmgnt/entity/Owner.java (hunk 1 above)
Comment (Samrudhsaish) [reply src/main/java/com/acro/dev/propmgnt/entity/Owner.java]: Sure Rani

**Review by rambabu — CHANGES_REQUESTED** [review]

File: src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+        if (owner.isPresent()) {
+            Owner ownerOne = owner.get();
+            address.setId(addressRequest.getAddressId());
+            address.setLineOne(addressRequest.getLineOne());
+            address.setLineTwo(addressRequest.getLineTwo());
+            address.setZipcode(addressRequest.getZipcode());
+            address.setCity(addressRequest.getCity());
+            address.setState(addressRequest.getState());
+            address.setType(addressRequest.getType());
+            Address addressOne = addressRepository.save(address);
```
Comment (Ranianirudh) [inline src/main/java/com/acro/dev/propmgnt/service/PropertyServiceImpl.java]: Address Details not provided here. May be we should fetch the address here.

File: src/main/java/com/acro/dev/propmgnt/responsemethod/CommonResponse.java [hunk 32]
```diff
@@ -0,0 +1,50 @@
+package com.acro.dev.propmgnt.responsemethod;
+
+import com.acro.dev.propmgnt.entity.Address;
+import com.acro.dev.propmgnt.entity.Owner;
+import com.acro.dev.propmgnt.entity.Property;
+import com.acro.dev.propmgnt.response.AddressResponse;
+import com.acro.dev.propmgnt.response.OwnerResponse;
+import com.acro.dev.propmgnt.response.PropertyResponse;
+
```
Comment (Ranianirudh) [inline src/main/java/com/acro/dev/propmgnt/responsemethod/CommonResponse.java:11]: This should be marked as a @Component

**Review by Ranianirudh — CHANGES_REQUESTED** [review]

**Review by rambabu — APPROVED** [review]

**Review by Ranianirudh — APPROVED** [review]

**Pull request merged**